File: dotnet/UrlShortener.Core/Factories/DatabaseFactory.cs
using Microsoft.Extensions.Configuration;
using UrlShortener.Core.Interfaces;
using UrlShortener.Core.Database;

namespace UrlShortener.Core.Factories;

public static class DatabaseFactory
{
    public static IUrlDatabase CreateDatabase(IConfiguration configuration)
    {
        var dbType = configuration["DB_TYPE"] ?? configuration["DbType"] ?? "sqlite";

        return dbType.ToLowerInvariant() switch
        {
            "sqlite" => new SqliteDatabase(
                configuration["DB_PATH"] ?? configuration["DbPath"] ?? "./urls.db"
            ),
            "sqlserver" => new SqlServerDatabase(
                GetRequiredConnectionString(configuration, "SQL Server")
            ),
            "postgresql" => new PostgreSqlDatabase(
                GetRequiredConnectionString(configuration, "PostgreSQL")
            ),
            "cosmosdb" => CreateCosmosDatabase(configuration),
            "azurestorage" => CreateAzureStorageDatabase(configuration),
            _ => throw new NotSupportedException($"Unsupported database type: {dbType}. Supported types: sqlite, sqlserver, postgresql, cosmosdb, azurestorage")
        };
    }

    private static string GetRequiredConnectionString(IConfiguration configuration, string dbType)
    {
        var connectionString = configuration["DB_CONNECTION_STRING"] ?? 
                             configuration["DbConnectionString"] ?? 
                             configuration.GetConnectionString("DefaultConnection");
        
        if (string.IsNullOrEmpty(connectionString))
        {
            throw new InvalidOperationException($"DB_CONNECTION_STRING is required for {dbType}");
        }
        
        return connectionString;
    }

    private static CosmosDbDatabase CreateCosmosDatabase(IConfiguration configuration)
    {
        var connectionString = configuration["DB_CONNECTION_STRING"] ?? configuration["DbConnectionString"];
        
        if (!string.IsNullOrEmpty(connectionString))
        {
            // Use connection string format
            return new CosmosDbDatabase(
                connectionString,
                configuration["COSMOSDB_DATABASE"] ?? configuration["CosmosDbDatabase"] ?? "urlshortener",
                configuration["COSMOSDB_CONTAINER"] ?? configuration["CosmosDbContainer"] ?? "urls"
            );
        }

        // Use legacy individual parameters
        var endpoint = configuration["COSMOS_DB_ENDPOINT"] ?? configuration["CosmosDbEndpoint"];
        var key = configuration["COSMOS_DB_KEY"] ?? configuration["CosmosDbKey"];
        
        if (string.IsNullOrEmpty(endpoint) || string.IsNullOrEmpty(key))
        {
            throw new InvalidOperationException("Either DB_CONNECTION_STRING or both COSMOS_DB_ENDPOINT and COSMOS_DB_KEY are required for Cosmos DB");
        }

        return new CosmosDbDatabase(
            endpoint,
            key,
            configuration["COSMOSDB_DATABASE"] ?? configuration["CosmosDbDatabase"] ?? "urlshortener",
            configuration["COSMOSDB_CONTAINER"] ?? configuration["CosmosDbContainer"] ?? "urls"
        );
    }

    private static AzureStorageTablesDatabase CreateAzureStorageDatabase(IConfiguration configuration)
    {
        var accountName = configuration["AZURE_STORAGE_ACCOUNT_NAME"] ?? configuration["AzureStorageAccountName"];
        var accountKey = configuration["AZURE_STORAGE_ACCOUNT_KEY"] ?? configuration["AzureStorageAccountKey"];
        
        if (string.IsNullOrEmpty(accountName) || string.IsNullOrEmpty(accountKey))
        {
            throw new InvalidOperationException("AZURE_STORAGE_ACCOUNT_NAME and AZURE_STORAGE_ACCOUNT_KEY are required for Azure Storage Tables");
        }

        return new AzureStorageTablesDatabase(
            accountName,
            accountKey,
            configuration["AZURE_STORAGE_TABLE_NAME"] ?? configuration["AzureStorageTableName"] ?? "urls"
        );
    }
}

File: node/function/src/db/DatabaseFactory.js
const SQLiteDatabase = require('./SQLiteDatabase');
const SQLServerDatabase = require('./SQLServerDatabase');
const PostgreSQLDatabase = require('./PostgreSQLDatabase');
const CosmosDBDatabase = require('./CosmosDBDatabase');
const AzureStorageTablesDatabase = require('./AzureStorageTablesDatabase');

class DatabaseFactory {
  static createDatabase() {
    const dbType = process.env.DB_TYPE || 'sqlite';
    
    switch (dbType.toLowerCase()) {
      case 'sqlite':
        return new SQLiteDatabase(process.env.DB_PATH || './urls.db');
      
      case 'sqlserver':
        if (!process.env.DB_CONNECTION_STRING) {
          throw new Error('DB_CONNECTION_STRING is required for SQL Server');
        }
        return new SQLServerDatabase(process.env.DB_CONNECTION_STRING);
      
      case 'postgresql':
        if (!process.env.DB_CONNECTION_STRING) {
          throw new Error('DB_CONNECTION_STRING is required for PostgreSQL');
        }
        return new PostgreSQLDatabase(process.env.DB_CONNECTION_STRING);
      
      case 'cosmosdb':
        if (!process.env.DB_CONNECTION_STRING && (!process.env.COSMOS_DB_ENDPOINT || !process.env.COSMOS_DB_KEY)) {
          throw new Error('Either DB_CONNECTION_STRING or both COSMOS_DB_ENDPOINT and COSMOS_DB_KEY are required for Cosmos DB');
        }
        
        if (process.env.DB_CONNECTION_STRING) {
          // Use connection string format
          return new CosmosDBDatabase(
            process.env.DB_CONNECTION_STRING,
            process.env.COSMOSDB_DATABASE || process.env.COSMOS_DB_DATABASE || 'urlshortener',
            process.env.COSMOSDB_CONTAINER || process.env.COSMOS_DB_CONTAINER || 'urls'
          );
        } else {
          // Use legacy individual parameters
          return new CosmosDBDatabase(
            process.env.COSMOS_DB_ENDPOINT,
            process.env.COSMOS_DB_KEY,
            process.env.COSMOSDB_DATABASE || process.env.COSMOS_DB_DATABASE || 'urlshortener',
            process.env.COSMOSDB_CONTAINER || process.env.COSMOS_DB_CONTAINER || 'urls'
          );
        }
      
      case 'azurestorage':
        if (!process.env.AZURE_STORAGE_ACCOUNT_NAME || !process.env.AZURE_STORAGE_ACCOUNT_KEY) {
          throw new Error('AZURE_STORAGE_ACCOUNT_NAME and AZURE_STORAGE_ACCOUNT_KEY are required for Azure Storage Tables');
        }
        return new AzureStorageTablesDatabase(
          process.env.AZURE_STORAGE_ACCOUNT_NAME,
          process.env.AZURE_STORAGE_ACCOUNT_KEY,
          process.env.AZURE_STORAGE_TABLE_NAME || 'urls'
        );
      
      default:
        throw new Error(`Unsupported database type: ${dbType}. Supported types: sqlite, sqlserver, postgresql, cosmosdb, azurestorage`);
    }
  }
}

module.exports = DatabaseFactory;
